### PR TITLE
py : add Gemma conversion from HF models

### PR DESCRIPTION
**Pull request by ggerganov**
```bash
# gemma-2b
python3 convert-hf-to-gguf.py ~/Data/huggingface/gemma-2b/ --outfile models/gemma-2b/ggml-model-f16.gguf --outtype f16

# gemma-7b
python3 convert-hf-to-gguf.py ~/Data/huggingface/gemma-7b/ --outfile models/gemma-7b/ggml-model-f16.gguf --outtype f16
```

### CODE REVIEW

**Comment by twoxfh** [comment]
I successfully created a 2b GGUF and loaded model with the server in master. Thanks!

File: convert-hf-to-gguf.py [hunk 1]
```diff
@@ -218,6 +218,8 @@ def from_model_architecture(model_architecture):
             return BertModel
         if model_architecture == "NomicBertModel":
             return NomicBertModel
+        if model_architecture in "GemmaForCausalLM":
```
Comment (akx) [inline convert-hf-to-gguf.py:221]: Shouldn't this be
```suggestion
        if model_architecture == "GemmaForCausalLM":
```
unless you actually do mean to do a substring search here?

File: convert-hf-to-gguf.py [hunk 2]
```diff
@@ -277,6 +279,8 @@ def _get_model_architecture(self) -> gguf.MODEL_ARCH:
             return gguf.MODEL_ARCH.BERT
         if arch == "NomicBertModel":
             return gguf.MODEL_ARCH.NOMIC_BERT
+        if arch in "GemmaForCausalLM":
```
Comment (akx) [inline convert-hf-to-gguf.py:282]: Ditto here?
```suggestion
        if arch == "GemmaForCausalLM":
```

**Review by Yefori-Go — APPROVED** [review]
Works well for me

**Comment by postmasters** [comment]
I notice that the HF `config.json` says there are 256000 tokens. But the embedding layer is 256128 x d_model. Not sure if there would be latent issues later.

File: convert-hf-to-gguf.py [hunk 3]
```diff
@@ -1785,6 +1789,62 @@ def get_tensors(self):
...
+        self.gguf_writer.add_key_length(hparams["head_dim"])
+        self.gguf_writer.add_value_length(hparams["head_dim"])
+
+    def write_tensors(self):
+        block_count = self.hparams.get("n_layers", self.hparams.get("num_hidden_layers", self.hparams.get("n_layer")))
+        tensor_map = gguf.get_tensor_name_map(self.model_arch, block_count)
+
+        for name, data_torch in self.get_tensors():
+            # ref: https://github.com/huggingface/transformers/blob/fc37f38915372c15992b540dfcbbe00a916d4fc6/src/transformers/models/gemma/modeling_gemma.py#L89
+            if name.endswith(("norm.weight")):
```
Comment (cebtenzzre) [inline convert-hf-to-gguf.py:1817]:
```suggestion
            if name.endswith("norm.weight"):
```

**Comment by Ronnie-Leon76** [comment]
image
I have tried quantizing a fine-tuned gemma-7b model that was loaded as 4-bit but I get the error: ``` Can not map tensor 'model.layers.0.mlp.down_proj.weight.absmax'``` above. @ggerganov I'll appreciate it if you help me resolve this issue.

**Comment by diogo-garcia** [comment]
> image I have tried quantizing a fine-tuned gemma-7b model that was loaded as 4-bit but I get the error: ` Can not map tensor 'model.layers.0.mlp.down_proj.weight.absmax'` above. @ggerganov I'll appreciate it if you help me resolve this issue.

I recently downloaded the Meta-Llama-3-8b model from huggingface and attempted to convert it to GGUF format using the following command: 

`python3 /root/llama.cpp/convert-hf-to-gguf.py /root/models/meta-llama-3-8b --outfile /root/models/meta-llama-3-8b.gguf --outtype f32`

However, I encountered the same error. The specific error message I received was:

>   File "/root/llama.cpp/convert-hf-to-gguf.py", line 182, in map_tensor_name
>     raise ValueError(f"Can not map tensor {name!r} {try_suffixes} {self.tensor_map}")
> ValueError: Can not map tensor '**model.layers.0.mlp.down_proj.weight.absmax**'


I also tried modifying the function map_tensor_name in the script to change the suffixes from .weight to .weight_map, which resolved the previous error, but now I am getting a new error:

>   File "/root/llama.cpp/convert-hf-to-gguf.py", line 182, in map_tensor_name
>     raise ValueError(f"Can not map tensor {name!r} {try_suffixes} {self.tensor_map}")
> ValueError: Can not map tensor '**model.embed_tokens.weight**'

The reason I made this change is that the structure of **model.safetensors.index.json** indicates that the weights are mapped with suffixes like .weight_map. Here is an excerpt from the JSON file for reference:

> {
>   "metadata": {
>     "total_size": 6027779904
>   },
>   "**weight_map**": {
>     "lm_head.weight": "model-00002-of-00002.safetensors",
>     "model.embed_tokens.weight": "model-00001-of-00002.safetensors",
>     "model.layers.0.input_layernorm.weight": "model-00001-of-00002.safetensors",
>     "model.layers.0.mlp.down_proj.weight": "model-00001-of-00002.safetensors",
>     "model.layers.0.mlp.down_proj.weight.absmax": "model-00001-of-00002.safetensors",
>     "model.layers.0.mlp.down_proj.weight.quant_map": "model-00001-of-00002.safetensors",
>     ...

Could anyone please provide guidance on how to properly map these tensors, or if there's a different approach needed for this conversion? Thank you!

**Pull request merged**